### PR TITLE
[12.0][ADD] base_tier_validation_helper

### PR DESCRIPTION
**Pull request by ps-tubtim**
This module helps to create, edit and delete many tier definitions.

### CODE REVIEW

**Comment by LoisRForgeFlow** [comment]
Interesting idea. Some question on the approach:
* why don't you implement it as a wizard?
* can't it be done without the dependency to tier_validation_formula?

**Comment by kittiu** [comment]
@lreficent 

Currently in base_tier_validation, there is no header model that can group tiers (which I do agree on the base simplicity). So, we like to propose this as separated set of addons.

1. To do header model for ease of maintenance
2. To make it simple to do different kind of complex use cases (1 addon per new use case, amount level is just one of them)
3. For each model, we could also have different set of tier validation, based on different department, etc.

> why don't you implement it as a wizard?

As per mentioned above. Do you think it is reasonable?

> can't it be done without the dependency to tier_validation_formula?

@ps-tubtim I do agree with @lreficent I think we can make it very generic like this,

- base_tier_validation_helper --> nothing more than the header model link to the needed model, action buttons (with no implementation yet). No need to inherit formula.
- base_tier_validation_helper_amount_level --> this is the implement for the use case that you are working on (validation by amount level). This will inherit formula.

In the future, we can have more use case on this helper, which may or may not need formula, i.e.,
- base_tier_validation_helper_xxx

What do you think?
Also, please feel free to comment on naming convention, we don't have a good name yet.

**Comment by kittiu** [comment]
@ps-tubtim something you may consider,

* As we split, and base_tier_validation_helper_amount_level is just one implementation. We have helper_type, as selection field.
* To ease user understanding, we can add detailed explaination and examples, in separated "Help" tab. Help in mis_builder is good example, https://github.com/OCA/mis-builder/blob/10.0/mis_builder/views/mis_report.xml#L111

**Comment by kittiu** [comment]
@lreficent if you don't mind, may be @ps-tubtim can continue. Should you have more input, do let us know anytime.

**Comment by LoisRForgeFlow** [comment]
@kittiu Thanks for the detailed response and sorry for my late answer.

> As per mentioned above. Do you think it is reasonable?

Yes, I agree that it can be good to keep that record to ease management and maintenance of models with a lot of tiers defined.

> base_tier_validation_helper --> nothing more than the header model link to the needed model, action buttons (with no implementation yet). No need to inherit formula.

I like this idea, as formula extension should be avoided as much as possible to reduce security risks (quite remote risk though, but we have to be aware). In this module we may not provide any logic, just as you say, a header to easily manipulate all the tier definitions for a model. So, you should be able to edit the domain in that view per each line.

> base_tier_validation_helper_amount_level --> this is the implement for the use case that you are working on (validation by amount level). This will inherit formula.

Once formula is inhereted I would not split modules more. New implementations and logics to manage the tiers could just extend this second module.

> Also, please feel free to comment on naming convention, we don't have a good name yet.

I don't have any problem with the name!

In summary my only suggestion: base_tier_validation_helper and base_tier_validation_helper_formula would be enough segregation. Each one of them can contain different logic to handle the tiers, logic with no need to use formula goes to base_tier_validation_helper and logic using formula goes to base_tier_validation_helper_formula

What do you think?

Thanks for your contributions and ideas :wink:

**Comment by kittiu** [comment]
@lreficent thanks for your reply. I do agree with most, just to add some thing further,

What I agree that, only 2 module is enough,
1. base_tier_validation_helper --> which is very basic.
2. base_tier_validation_helper_formula --> the implementation goes here.

I want to add that `base_tier_validation_helper_formula` could be the end module for formula use. Today we only have 1 use case. In the future if we have more, we can just enhance it here as another type.

This is to avoid too many modules which can be confusing. Having said that, for now, your advise is still valid. :)

@ps-tubtim agreed? anything you want to add?

**Comment by LoisRForgeFlow** [comment]
@kittiu Yes, I agree. Feel free to proceed with the changes!

File: base_tier_validation_helper/models/tier_definition_helper.py [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+    _name = 'tier.definition.helper.line'
+    _description = 'Tier Definition Helper Line'
+
+    parent_id = fields.Many2one(
+        comodel_name='tier.definition.helper',
+        index=True,
+        required=True,
+        ondelete='cascade'
+    )
+    approval_level = fields.Integer(
```
Comment (HviorForgeFlow) [inline base_tier_validation_helper/models/tier_definition_helper.py:148]: I couldn't figure out how this variable worked until I see the code, please add a help explaining that levels go from lowest one 1 to 100, I don't see any restriction to not use a number higher to 100.

File: base_tier_validation_helper/models/tier_definition_helper.py [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+        string="Reviewer",
+    )
+    reviewer_group_id = fields.Many2one(
+        comodel_name="res.groups",
+        string="Reviewer group",
+    )
+    definition_domain = fields.Char(
+        string='Domain',
+        help='Expressions can be any valid python expressions.'
+             '(eg ["|", ["id", "=", 1], ["id", "=", "2"]])',
```
Comment (HviorForgeFlow) [inline base_tier_validation_helper/models/tier_definition_helper.py]:
```suggestion
             '(eg ["|", ("id", "=", 1), ("id", "=", 2)])',
```
Should it  be this way?

File: base_tier_validation_helper/models/tier_definition_helper.py [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+
+class TierDefinitionHelper(models.Model):
+    _name = 'tier.definition.helper'
+    _description = 'Tier Definition Helper'
+
+    @api.model
+    def _get_default_name(self):
+        return _('New')
+
+    name = fields.Char(
```
Comment (kittiu) [inline base_tier_validation_helper/models/tier_definition_helper.py:12]: I think we should not allow name duplication for the helper (it will make no sense to have same name).
- Please add sql unique constraints
- When copy, should have (Copy) suffix

File: base_tier_validation_helper/models/tier_definition_helper.py [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+        string='Description',
+        required=True,
+        default=lambda self: self._get_default_name(),
```
Comment (kittiu) [inline base_tier_validation_helper/models/tier_definition_helper.py]:
```suggestion
        default='New',
```
I just think it will be less code for almost the same thing.

File: base_tier_validation_helper/models/tier_definition_helper.py [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+        inverse_name='parent_id',
+        string='Lines',
+        readonly=True,
+        states={'draft': [('readonly', False)]},
+    )
+
+    @api.multi
+    @api.constrains('line_ids')
+    def _check_line_ids(self):
+        if self.state == 'done' and \
```
Comment (kittiu) [inline base_tier_validation_helper/models/tier_definition_helper.py]: As you use @api.multi, we need for loop on self.

Comment (kittiu) [reply base_tier_validation_helper/models/tier_definition_helper.py]: But why do we need this constraint? line_ids already readonly on state=='done' isn't it?

File: base_tier_validation_helper/models/tier_definition_helper.py [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+        models = self.env['tier.definition']._get_tier_validation_model_names()
+        return {'domain': {'model_id': [('model', 'in', models)]}}
+
+    @api.multi
+    def button_draft(self):
+        self.write({'state': 'draft'})
+
+    @api.multi
+    def button_create_tier(self):
+        if not self.line_ids:
```
Comment (kittiu) [inline base_tier_validation_helper/models/tier_definition_helper.py]: We use @api.multi, right? Please loop.
Otherwise, use self.mapped('line_ids')

File: base_tier_validation_helper/models/tier_definition_helper.py [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+            raise ValidationError(_('Please add at least one line.'))
+        tier_definition_env = self.env['tier.definition']
```
Comment (kittiu) [inline base_tier_validation_helper/models/tier_definition_helper.py]: Naming convension, use Camel case for class.
```suggestion
        TierDefinition = self.env['tier.definition']
```

File: base_tier_validation_helper/models/tier_definition_helper.py [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+        for rec in self.line_ids:
```
Comment (kittiu) [inline base_tier_validation_helper/models/tier_definition_helper.py]: Self can be multi, please use for loop.

File: base_tier_validation_helper/models/tier_definition_helper.py [hunk 9]
```diff
@@ -0,0 +1,230 @@
...
+            if not rec.tier_definition_id:
+                tier_definition = tier_definition_env.create(
+                    rec._prepare_tier_definition())
+                rec.tier_definition_id = tier_definition
```
Comment (kittiu) [inline base_tier_validation_helper/models/tier_definition_helper.py]: Can be done in one line to shorten, i.e.,

```
res.tier_definition_id = TierDefinition.create(rec._prepare_tier_definition())
```

File: base_tier_validation_helper/models/tier_definition_helper.py [hunk 10]
```diff
@@ -0,0 +1,230 @@
...
+        self.write({'state': 'done'})
+
+    @api.multi
+    def button_delete_tier(self):
+        self.line_ids.mapped('tier_definition_id').unlink()
```
Comment (kittiu) [inline base_tier_validation_helper/models/tier_definition_helper.py]:
```suggestion
        self.mapped('line_ids').mapped('tier_definition_id').unlink()
```

File: base_tier_validation_helper/models/tier_definition_helper.py [hunk 11]
```diff
@@ -0,0 +1,230 @@
...
+        self.write({'state': 'cancel'})
+
+
+class TierDefinitionHelperLine(models.Model):
+    _name = 'tier.definition.helper.line'
+    _description = 'Tier Definition Helper Line'
+
+    parent_id = fields.Many2one(
```
Comment (kittiu) [inline base_tier_validation_helper/models/tier_definition_helper.py]: Naming Convention
(parent_id normally means same object but have hierarchy)
```suggestion
    helper_id = fields.Many2one(
```

File: base_tier_validation_helper/models/tier_definition_helper.py [hunk 12]
```diff
@@ -0,0 +1,230 @@
...
+
+    @api.multi
+    def button_delete_tier(self):
+        self.line_ids.mapped('tier_definition_id').unlink()
+        self.write({'state': 'cancel'})
+
+
+class TierDefinitionHelperLine(models.Model):
+    _name = 'tier.definition.helper.line'
+    _description = 'Tier Definition Helper Line'
```
Comment (kittiu) [inline base_tier_validation_helper/models/tier_definition_helper.py:135]: Should it _order by approval_level?

File: base_tier_validation_helper/models/tier_definition_helper.py [hunk 13]
```diff
@@ -0,0 +1,230 @@
...
+        index=True,
+        required=True,
+        ondelete='cascade'
+    )
+    approval_level = fields.Integer(
+        string='Level',
+        default=1,
+        required=True,
+    )
+    name = fields.Char(
```
Comment (kittiu) [inline base_tier_validation_helper/models/tier_definition_helper.py:153]: name should be requried.

File: base_tier_validation_helper/models/tier_definition_helper.py [hunk 14]
```diff
@@ -0,0 +1,230 @@
...
+        string='Name',
+        default='New Tier Definition',
+    )
+    reviewer_id = fields.Many2one(
+        comodel_name="res.users",
```
Comment (kittiu) [inline base_tier_validation_helper/models/tier_definition_helper.py]: Double quote or single quote, please choose one.

File: base_tier_validation_helper/models/tier_definition_helper.py [hunk 15]
```diff
@@ -0,0 +1,230 @@
...
+        string="Reviewer",
+    )
+    reviewer_group_id = fields.Many2one(
+        comodel_name="res.groups",
+        string="Reviewer group",
+    )
+    definition_domain = fields.Char(
```
Comment (kittiu) [inline base_tier_validation_helper/models/tier_definition_helper.py]: I think domain should be in header, not in line.

File: base_tier_validation_helper/models/tier_definition_helper.py [hunk 16]
```diff
@@ -0,0 +1,230 @@
...
+            'review_type': self.parent_id.review_type,
+            'definition_type': self.parent_id.definition_type,
+            'notify_on_create': self.parent_id.notify_on_create,
+            'has_comment': self.parent_id.has_comment,
+            'approve_sequence': self.parent_id.approve_sequence,
+        }
+        # add name
+        if self.parent_id.auto_name:
+            tier_definition['name'] = self.parent_id.model_id.name +\
+                ' (' + str(self.approval_level) + ')'
```
Comment (kittiu) [inline base_tier_validation_helper/models/tier_definition_helper.py]: Please always do not use '+' for sting.
```
name = '%s (%s)' % (var1, var2)
```

Comment (kittiu) [reply base_tier_validation_helper/models/tier_definition_helper.py]: Anyway, I think auto_name is not needed (let use make name himself is ok).

File: base_tier_validation_helper/models/tier_definition_helper.py [hunk 17]
```diff
@@ -0,0 +1,230 @@
...
+            'model_id': self.parent_id.model_id.id,
+            'sequence': 100 - self.approval_level,
+            'review_type': self.parent_id.review_type,
+            'definition_type': self.parent_id.definition_type,
+            'notify_on_create': self.parent_id.notify_on_create,
+            'has_comment': self.parent_id.has_comment,
+            'approve_sequence': self.parent_id.approve_sequence,
+        }
+        # add name
+        if self.parent_id.auto_name:
```
Comment (kittiu) [inline base_tier_validation_helper/models/tier_definition_helper.py]: multi or ensure_one() is needed.

File: base_tier_validation_helper/models/tier_definition_helper.py [hunk 18]
```diff
@@ -0,0 +1,230 @@
...
+        if self.parent_id.definition_type == 'domain':
+            tier_definition['definition_domain'] = self.definition_domain
+        return tier_definition
+
+    @api.onchange('approval_level')
+    def _onchange_approval_level(self):
+        if self.tier_definition_id:
+            tier_definition = self.env['tier.definition'].search(
+                [('id', '=', self.tier_definition_id.id)], limit=1)
+            tier_definition.write({'sequence': 100 - self.approval_level})
```
Comment (kittiu) [inline base_tier_validation_helper/models/tier_definition_helper.py]: What number 100 is from? What if approval_level is > 100?
Or can we find max(approval_level) and make the number +1, etc...

File: base_tier_validation_helper/models/tier_definition_helper.py [hunk 19]
```diff
@@ -0,0 +1,230 @@
...
+
+    @api.onchange('name')
+    def _onchange_name(self):
+        if self.tier_definition_id:
```
Comment (kittiu) [inline base_tier_validation_helper/models/tier_definition_helper.py]: What is this for? Do we still need it?

File: base_tier_validation_helper/models/tier_definition_helper.py (hunk 5 above)
Comment (kittiu) [reply base_tier_validation_helper/models/tier_definition_helper.py]: We can use in view, create="0" and reduce state to only Draft -> Done.

File: base_tier_validation_helper/models/tier_definition_helper.py (hunk 1 above)
Comment (kittiu) [reply base_tier_validation_helper/models/tier_definition_helper.py:148]: Let's use max(...) to find the right number?

File: base_tier_validation_helper/models/tier_definition_helper.py (hunk 19 above)
Comment (kittiu) [reply base_tier_validation_helper/models/tier_definition_helper.py]: I think many of the onchange can be removed, if we only allowed 2 states, Draft -> Done. When set to draft, all lines will be deleted and restart all over again.
No need to edit line when it is active.

**Comment by kittiu** [comment]
@ps-tubtim 
And because the intention of this module is to ease user (otherwise, user can use standard one). I think the "Domain" should be defined at the header level (then to cascade to all tiers).

File: base_tier_validation_helper/models/tier_definition_helper.py (hunk 3 above)
Comment (ps-tubtim) [reply base_tier_validation_helper/models/tier_definition_helper.py:12]: Fixed

File: base_tier_validation_helper/models/tier_definition_helper.py (hunk 4 above)
Comment (ps-tubtim) [reply base_tier_validation_helper/models/tier_definition_helper.py]: Fixed

File: base_tier_validation_helper/models/tier_definition_helper.py (hunk 5 above)
Comment (ps-tubtim) [reply base_tier_validation_helper/models/tier_definition_helper.py]: Fixed

File: base_tier_validation_helper/models/tier_definition_helper.py (hunk 6 above)
Comment (ps-tubtim) [reply base_tier_validation_helper/models/tier_definition_helper.py]: Fixed

File: base_tier_validation_helper/models/tier_definition_helper.py (hunk 7 above)
Comment (ps-tubtim) [reply base_tier_validation_helper/models/tier_definition_helper.py]: Edited

File: base_tier_validation_helper/models/tier_definition_helper.py (hunk 8 above)
Comment (ps-tubtim) [reply base_tier_validation_helper/models/tier_definition_helper.py]: Fixed

File: base_tier_validation_helper/models/tier_definition_helper.py (hunk 9 above)
Comment (ps-tubtim) [reply base_tier_validation_helper/models/tier_definition_helper.py]: Fixed

File: base_tier_validation_helper/models/tier_definition_helper.py (hunk 10 above)
Comment (ps-tubtim) [reply base_tier_validation_helper/models/tier_definition_helper.py]: Fixed

File: base_tier_validation_helper/models/tier_definition_helper.py (hunk 11 above)
Comment (ps-tubtim) [reply base_tier_validation_helper/models/tier_definition_helper.py]: Edited

File: base_tier_validation_helper/models/tier_definition_helper.py (hunk 12 above)
Comment (ps-tubtim) [reply base_tier_validation_helper/models/tier_definition_helper.py:135]: Fixed

File: base_tier_validation_helper/models/tier_definition_helper.py (hunk 13 above)
Comment (ps-tubtim) [reply base_tier_validation_helper/models/tier_definition_helper.py:153]: Fixed

File: base_tier_validation_helper/models/tier_definition_helper.py (hunk 14 above)
Comment (ps-tubtim) [reply base_tier_validation_helper/models/tier_definition_helper.py]: Fixed

File: base_tier_validation_helper/models/tier_definition_helper.py (hunk 15 above)
Comment (ps-tubtim) [reply base_tier_validation_helper/models/tier_definition_helper.py]: Fixed

File: base_tier_validation_helper/models/tier_definition_helper.py (hunk 1 above)
Comment (ps-tubtim) [reply base_tier_validation_helper/models/tier_definition_helper.py:148]: Fixed

File: base_tier_validation_helper/models/tier_definition_helper.py (hunk 2 above)
Comment (ps-tubtim) [reply base_tier_validation_helper/models/tier_definition_helper.py]: Edited

File: base_tier_validation_helper/models/tier_definition_helper.py (hunk 16 above)
Comment (ps-tubtim) [reply base_tier_validation_helper/models/tier_definition_helper.py]: Fixed

File: base_tier_validation_helper/models/tier_definition_helper.py (hunk 17 above)
Comment (ps-tubtim) [reply base_tier_validation_helper/models/tier_definition_helper.py]: Fixed

File: base_tier_validation_helper/models/tier_definition_helper.py (hunk 18 above)
Comment (ps-tubtim) [reply base_tier_validation_helper/models/tier_definition_helper.py]: Fixed

File: base_tier_validation_helper/models/tier_definition_helper.py (hunk 19 above)
Comment (ps-tubtim) [reply base_tier_validation_helper/models/tier_definition_helper.py]: Fixed

**Comment by ps-tubtim** [comment]
@kittiu @hveficent Thank you for review. Edited.

**Comment by LoisRForgeFlow** [comment]
@kittiu Is this ready?

**Comment by kittiu** [comment]
@LoisRForgeFlow we may have the real use case later on in v13 or v14. But for now, can we keep this pending (if no one request for it yet).
